src/modules/internal_combustion_engine_control/InternalCombustionEngineControl.cpp

```diff
@@ -0,0 +1,359 @@
+
+using namespace time_literals;
+
+namespace internal_combustion_engine_control
+{
+
+InternalCombustionEngineControl::InternalCombustionEngineControl() :
+	ModuleParams(nullptr),
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::lp_default)
+{
+	_internal_combustion_engine_control_pub.advertise();
+	_internal_combustion_engine_status_pub.advertise();
+}
+
+InternalCombustionEngineControl::~InternalCombustionEngineControl()
+{
+
+}
+
+int InternalCombustionEngineControl::task_spawn(int argc, char *argv[])
+{
+	InternalCombustionEngineControl *obj = new InternalCombustionEngineControl();
+
+	if (!obj) {
+		PX4_ERR("alloc failed");
+		return -1;
+	}
+
+	_object.store(obj);
+	_task_id = task_id_is_work_queue;
+
+	/* Schedule a cycle to start things. */
+	obj->start();
+
+	return 0;
+}
+
+void InternalCombustionEngineControl::start()
+{
+	ScheduleOnInterval(20_ms); // 50 Hz
+}
+
+void InternalCombustionEngineControl::Run()
+{
+	if (should_exit()) {
+		ScheduleClear();
+		exit_and_cleanup();
+	}
+
+	// check for parameter updates
+	if (_parameter_update_sub.updated()) {
+		// clear update
+		parameter_update_s pupdate;
+		_parameter_update_sub.copy(&pupdate);
+
+		// update parameters from storage
+		updateParams();
+		_throttle_control_slew_rate.setSlewRate(_param_ice_thr_slew.get());
+	}
+
+	UserOnOffRequest user_request = UserOnOffRequest::Keep; // todo: keep is not yet doing anything
+
+	manual_control_setpoint_s manual_control_setpoint;
+	_manual_control_setpoint_sub.copy(&manual_control_setpoint);
+
+	vehicle_status_s vehicle_status;
+	_vehicle_status_sub.copy(&vehicle_status);
+
+	actuator_motors_s actuator_motors;
+	_actuator_motors.copy(&actuator_motors);
+
+	const float throttle_in = actuator_motors.control[0];
+
+	switch (static_cast<ICESource>(_param_ice_on_source.get())) {
+	case ICESource::None:
+		user_request = UserOnOffRequest::Off;
+		break;
+
+	case ICESource::ArmingState:
+		user_request = vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED ? UserOnOffRequest::On :
+			       UserOnOffRequest::Off;
+		break;
+
+	case ICESource::Mavlink:
+		user_request = UserOnOffRequest::Off;
+		break;
+
+	case ICESource::Aux1:
+		user_request = manual_control_setpoint.aux1 > 0.5f ? UserOnOffRequest::On : UserOnOffRequest::Off;
+		break;
+
+	case ICESource::Aux2:
+		user_request = manual_control_setpoint.aux2 > 0.5f ? UserOnOffRequest::On : UserOnOffRequest::Off;
+		break;
+
+	case ICESource::Aux3:
+		user_request = manual_control_setpoint.aux3 > 0.5f ? UserOnOffRequest::On : UserOnOffRequest::Off;
+		break;
+
+	case ICESource::Aux4:
+		user_request = manual_control_setpoint.aux4 > 0.5f ? UserOnOffRequest::On : UserOnOffRequest::Off;
+		break;
+	}
+
+	internal_combustion_engine_control_s ice_control{};
+
+	switch (_state) {
+	case State::Stopped:
+
+		controlEngineStop(ice_control);
+
+		if (user_request == UserOnOffRequest::On && !_engine_tried_to_restart) {
+			_state = State::Starting;
+			instantiateEngineStart();
```

Comment:
```suggestion
			_state_start_time = now;
```
This method isn't needed anymore I would say.

And just store `now` at the beginning of the Run() function.